### PR TITLE
fix(core): Reconcile webhook triggers against stored state during publication (no-changelog)

### PR DESCRIPTION
**Pull request by mfsiega**
## Summary

When publishing a workflow, we register any webhooks that should be in the new version but aren't in the webhook table.

This closes a gap in error handling where we crash after advancing the workflow published version but before we finish registering webhooks. In this case, the diff against the old version wouldn't yield anything. We already do a diff against the live state for non-webhook triggers; this is analogous for webhooks.

This *doesn't* reconcile any 3rd-party state. That's left for a future improvement. If a node is missing one of multiple webhooks, we re-register the whole node. This should be safely idempotent.

## How to test

Added unit tests.

Manual: with the flag on, publish a workflow with a webhook trigger, delete its `webhook_entity` row to simulate a partial/missing registration, republish the same version, and confirm the webhook is re-registered.

## Related Linear tickets, Github issues, and Community forum posts

https://linear.app/n8n/issue/CAT-3536

## Review / Merge checklist

- [X] I have seen this code, I have run this code, and I take responsibility for this code.
- [X] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md))
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included.
- [ ] PR Labeled with `Backport to Beta`, `Backport to Stable`, or `Backport to v1` (if the PR is an urgent fix that needs to be backported)

🤖 PR Summary generated by AI

---

<a href="https://stagereview.app/n8n-io/n8n/pull/32895">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

<a href="https://cubic.dev/pr/n8n-io/n8n/pull/32895?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Review by tomi — DISMISSED** [review]
LGTM, just a comment and question

File: packages/cli/src/workflows/triggers/webhook-trigger-registrar.ts [hunk 1]
```diff
@@ -163,6 +153,75 @@ export class WebhookTriggerRegistrar {
...
+		const desiredWebhooks = this.getWebhookTriggers(workflow, additionalData).filter(
+			(webhookData) => desiredNodes.has(workflow.getNode(webhookData.node)?.id ?? ''),
+		);
+		if (desiredWebhooks.length === 0) {
+			return new Set();
+		}
+
+		const registeredKeys = new Set(
+			(await this.webhookService.getRegisteredWebhooks(workflow.id)).map((webhook) =>
+				this.webhookKey(webhook.method, webhook.webhookPath),
```
Comment (tomi) [inline packages/cli/src/workflows/triggers/webhook-trigger-registrar.ts]: I'm trying to think if this works in a scenario where a webhook node is delete within a workflow, and then readded in a different node, so the node name would be different. I guess it works? At least it sounds such an odd corner case that might not be be worth spending too much time on it

File: packages/cli/src/workflows/triggers/webhook-trigger-registrar.ts [hunk 2]
```diff
@@ -163,6 +153,75 @@ export class WebhookTriggerRegistrar {
...
+			method: webhookData.httpMethod,
+		});
+
+		this.normalizeWebhookPath(webhook, node.webhookId);
+
+		return webhook;
+	}
+
+	/** Identity of a webhook row: its `(method, path)` primary key. */
+	private webhookKey(method: string, webhookPath: string): string {
```
Comment (tomi) [inline packages/cli/src/workflows/triggers/webhook-trigger-registrar.ts]: nit: buildWebhookKey to align naming with the other method?

File: packages/cli/src/workflows/triggers/webhook-trigger-registrar.ts (hunk 1 above)
Comment (mfsiega) [reply packages/cli/src/workflows/triggers/webhook-trigger-registrar.ts]: I guess if the node were deleted, it would show up in the `toRemove` section and we wouldn't bump the workflow published version until we succeeded. Maybe there's a corner case where the nodes stay the same but an individual path is moved from one node to another?

I'm working on a follow-up PR that adds an e2e test here, so I can test that case.

File: packages/cli/src/workflows/triggers/webhook-trigger-registrar.ts (hunk 2 above)
Comment (mfsiega) [reply packages/cli/src/workflows/triggers/webhook-trigger-registrar.ts]: Done.

**Review by tomi — APPROVED** [review]
🚀
